src/serializer/utils.js

```diff
@@ -22,17 +30,41 @@ import type { AdditionalFunctionEffects } from "./types";
 
 /**
  * Get index property list length by searching array properties list for the max index key value plus 1.
+ * If tail elements are conditional, return the minimum length if an assignment to the length property
+ * can be avoided because of that. The boolean part of the result is a flag that indicates if the latter is true.
  */
-export function getSuggestedArrayLiteralLength(realm: Realm, val: ObjectValue): number {
+export function getSuggestedArrayLiteralLength(realm: Realm, val: ObjectValue): [number, boolean] {
   invariant(IsArray(realm, val));
 
-  let length = 0;
+  let minLength = 0,
+    maxLength = 0;
+  let actualLength;
   for (const key of val.properties.keys()) {
-    if (IsArrayIndex(realm, key) && Number(key) >= length) {
-      length = Number(key) + 1;
+    if (IsArrayIndex(realm, key) && Number(key) >= maxLength) {
+      let prevMax = maxLength;
+      maxLength = Number(key) + 1;
+      let elem = val._SafeGetDataPropertyValue(key);
+      if (!elem.mightHaveBeenDeleted()) minLength = maxLength;
+      else if (elem instanceof AbstractValue && elem.kind === "conditional") {
+        let maxLengthVal = new IntegralValue(realm, maxLength);
+        let [c, x, y] = elem.args;
+        if (x instanceof EmptyValue && !y.mightHaveBeenDeleted()) {
+          let prevActual = actualLength === undefined ? new IntegralValue(realm, prevMax) : actualLength;
+          actualLength = AbstractValue.createFromConditionalOp(realm, c, prevActual, maxLengthVal);
+        } else if (y instanceof EmptyValue && !x.mightHaveBeenDeleted()) {
+          let prevActual = actualLength === undefined ? new IntegralValue(realm, prevMax) : actualLength;
+          actualLength = AbstractValue.createFromConditionalOp(realm, c, maxLengthVal, prevActual);
+        } else {
+          actualLength = undefined;
+        }
+      }
     }
   }
-  return length;
+  if (maxLength > minLength && actualLength instanceof AbstractValue) {
+    let lengthVal = val._SafeGetDataPropertyValue("length");
+    if (lengthVal.equals(actualLength)) return [minLength, true];
```

Comment:
This .equals check seems very brittle to me. In the sense that I am not sure if without an SMT solver it will work for much more than the trivial test cases?

Reply:
The specific scenario that this enables is something like "loopvar < expressionWithSmallValuesDomainThatIsNotTop". Such expressions can be deterministically falsified using only the ValuesDomain. Of course, an SMT solver would be able to handle many more expressions, but my understanding from you was that the more limited use case shows up in actual code.